solr/core/src/java/org/apache/solr/util/circuitbreaker/MemoryCircuitBreaker.java

```diff
@@ -43,8 +46,34 @@ public class MemoryCircuitBreaker extends CircuitBreaker {
   private static final ThreadLocal<Long> seenMemory = ThreadLocal.withInitial(() -> 0L);
   private static final ThreadLocal<Long> allowedMemory = ThreadLocal.withInitial(() -> 0L);
 
+  /** Creates an instance which averages over 6 samples during last 30 seconds. */
   public MemoryCircuitBreaker() {
+    this(6, 5);
+  }
+
+  /**
+   * Constructor that allows override of sample interval for which the memory usage is fetched. This
+   * is provided for testing, not intended for general use.
+   *
+   * @param numSamples number of samples to calculate average for
+   * @param sampleInterval interval between each sample
+   */
+  protected MemoryCircuitBreaker(int numSamples, int sampleInterval) {
     super();
+    if (averagingMetricProvider == null || averagingMetricProvider.getRefcount() == 0) {
+      averagingMetricProvider =
```

Comment:
Wondering how this would work if two threads concurrently instantiated a circuit breaker e.g. two shards in the same JVM. Might an alternative be to

```
private static RefCounted<AveragingMetricProvider> averagingMetricProvider = ...
```

initialise always with the test constructor doing a replacement?

Reply:
It's my first attempt using `RefCounted` API, so I may be holding it wrong.

Reply:
No worries, I haven't yet used the `RefCounted` API either, perhaps something for me to learn about next week.

As an aside, just noticed the TODO in the `isTripped()` comments below and it sounds like the changes in this PR will have an effect similar to what is described there, maybe.

Reply:
Yes, let's remove that TODO, as averaging will have about the same effect. Also the isTripped method is now almost "free", as it just pulls the cached average.

Reply:
So I made both constructor and close methods synchronized to safeguard against this.